File: LC/221.maximal-square.py
class Solution(object):
    def maximalSquare(self, matrix):
        # Handle edge case
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return 0
        
        # Record length of edge of the max-sized square
        square_max = int(matrix[0][0])

        # The dp array is named 'arr'
        arr = []

        # Set initial values of arr
        for i in range(len(matrix[0])):
            arr.append(int(matrix[0][i]))
            if arr[-1] > square_max:
                square_max = arr[-1]        
        
        # Iterate through every value of the matrix
        for i in range(1, len(matrix)):
              
            # Set initial variables before inner loop
            # Temp tracks upper left square size, arr[j] the upper suqare, and arr[j-1] the left square
            temp = arr[0]
            arr[0] = int(matrix[i][0])
            if arr[0] > square_max:
                square_max = arr[0]

            # Updates arr with previous values on every iteration
            for j in range(1, len(matrix[0])):
                temp_new = arr[j]
                if matrix[i][j] == "1":
                    arr[j] = min(arr[j], arr[j-1], temp) +1
                else:
                    arr[j] = 0
                
                # Reset temp, temp tracks the upper_left square
                temp = temp_new
                if arr[j] > square_max:
                    square_max = arr[j]

        return square_max*square_max


#        """
#        :type matrix: List[List[str]]
#        :rtype: int
#        """

z = Solution()
matrix = [["1","0","1","0","0"],
          ["1","0","1","1","1"],
          ["1","1","1","1","1"],
          ["1","0","0","1","0"]]

_matrix = [
    ["1","0","1","1","1"],
    ["0","1","0","1","0"],
    ["1","1","0","1","1"],
    ["1","1","0","1","1"],
    ["0","1","1","1","1"]]

_matrix = [
    ["0","0","0","0","0"],
    ["1","0","0","0","0"],
    ["0","0","0","0","0"],
    ["0","0","0","0","0"]
]

_matrix = [
    ["0", "0"]
]
print(z.maximalSquare(matrix))

File: Leetcode/221-maximal-square.cpp
#include <iostream>
#include <vector>


using namespace std;

class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();
        vector<vector<int>> dp(n, vector<int>(m, 0));

        
        int max_square_size = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (matrix[i][j] == 1) {
                    if (i > 0 && j > 0) {
                        dp[i][j] = min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]) + 1;
                    } else {
                        dp[i][j] = 1;
                    }
                }
                max_square_size = max(max_square_size, dp[i][j]);
            }
        }
        return max_square_size;
    }
};

int main() {
    Solution sol;
    vector<vector<int>> matrix = {
        {"1","0","1","0","0"},
        {"1","0","1","1","1"},
        {"1","1","1","1","1"},
        {"1","0","0","1","0"}
    };
    cout << sol.maximalSquare(matrix) << endl;
}